cpp/tests/stream_compaction/distinct_count_tests.cpp

```diff
@@ -0,0 +1,310 @@
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  constexpr auto expected = 4;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::INCLUDE, nan_policy::NAN_IS_NULL));
+}
+
+TEST_F(DistinctCount, NansAsNullWithIgnoreNull)
+{
+  using T = float;
+
+  std::vector<T> input               = {1, 3, NAN, 70, 31};
+  std::vector<cudf::size_type> valid = {1, 1, 1, 0, 1};
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{input.begin(), input.end(), valid.begin()};
+
+  constexpr auto expected = 3;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_NULL));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_NULL));
+}
+
+TEST_F(DistinctCount, EmptyColumn)
+{
+  using T = float;
+
+  cudf::test::fixed_width_column_wrapper<T> input_col{};
+
+  constexpr auto expected = 0;
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_NULL));
+  EXPECT_EQ(expected,
+            cudf::distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_NULL));
+}
+
+TEST_F(DistinctCount, StringColumnWithNull)
+{
+  cudf::test::strings_column_wrapper input_col{
+    {"", "this", "is", "this", "This", "a", "column", "of", "the", "strings"},
+    {1, 1, 1, 1, 1, 1, 1, 1, 0, 1}};
+
+  cudf::size_type const expected =
+    (std::vector<std::string>{"", "this", "is", "This", "a", "column", "of", "strings"}).size();
+  EXPECT_EQ(
+    expected,
+    cudf::unordered_distinct_count(input_col, null_policy::EXCLUDE, nan_policy::NAN_IS_VALID));
+}
+
+TEST_F(DistinctCount, TableWithNull)
+{
+  cudf::test::fixed_width_column_wrapper<int32_t> col1{{5, 4, 3, 5, 8, 1, 4, 5, 0, 9, -1},
+                                                       {1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0}};
+  cudf::test::fixed_width_column_wrapper<int32_t> col2{{2, 2, 2, -1, 2, 1, 2, 0, 0, 9, -1},
+                                                       {1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0}};
+  cudf::table_view input{{col1, col2}};
+
+  EXPECT_EQ(8, cudf::unordered_distinct_count(input, null_equality::EQUAL));
+  EXPECT_EQ(10, cudf::unordered_distinct_count(input, null_equality::UNEQUAL));
+}
+
+TEST_F(DistinctCount, EmptyColumnedTable)
+{
+  std::vector<cudf::column_view> cols{};
+
+  cudf::table_view input(cols);
+
+  EXPECT_EQ(0, cudf::unordered_distinct_count(input, null_equality::EQUAL));
+  EXPECT_EQ(0, cudf::unordered_distinct_count(input, null_equality::UNEQUAL));
+  EXPECT_EQ(0, cudf::unordered_distinct_count(cudf::table_view{}, null_equality::EQUAL));
+  EXPECT_EQ(0, cudf::unordered_distinct_count(cudf::table_view{}, null_equality::UNEQUAL));
```

Comment:
Are the tests of `input` identical to the tests of the default-constructed `cudf::table_view{}`? It seems both would have an empty vector of columns.
```suggestion
  EXPECT_EQ(0, cudf::unordered_distinct_count(cudf::table_view{}, null_equality::EQUAL));
  EXPECT_EQ(0, cudf::unordered_distinct_count(cudf::table_view{}, null_equality::UNEQUAL));
```